### PR TITLE
CatLover assignment: Orestis Zinelis

### PR DESCRIPTION
**Pull request by OrestisZinelis**
# CatLover 

A React application for **cat lovers**. Display cat images, mark them as favorites and check the breed info!

## 📸 Screenshots

CatLovers-Cats
CatLovers-Breeds
CatLovers-Cats-Responsive
CatLovers-Breed-Modal
CatLovers-Cat-Modal

## Tech Stack

- React
- React Router
- TypeScript
- Vite
- Axios
- TanStack Query
- Tailwind CSS
- Material UI

## API

Get your API key from [The Cat API](https://developers.thecatapi.com/), then set `VITE_CAT_API_KEY` inside the `.env` file.

## Installation

1. Clone the repo

```sh
git clone https://github.com/OrestisZinelis/cat-lover
```

2. Navigate to the project directory e.g.

```sh
cd cat-lover
```

3. Install dependencies

```sh
yarn
```

or

```sh
npm install
```

4. Create a `.env` file in the root directory and add your API key:

`VITE_CAT_API_KEY=your_api_key_here`

## Development

Start the development server:

```sh
yarn dev
```

This will launch the app on http://localhost:5173

## Build

Build the project for production:

```sh
yarn build
```

This will generate the build files in the dist/ directory.

## Linting

Lint the project files using ESLint:

```sh
yarn lint
```

## Preview

Preview the built project:

```sh
yarn preview
```

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    open: false,
+    message: '',
+    severity: 'success'
+  })
+
+  const showAlert = useMemo(
+    () => (message: string, severity: 'success' | 'error') => {
+      setAlert({ open: true, message, severity })
+    },
+    [setAlert]
```
Comment (padelis) [inline src/App.tsx]: Is `setAlert` actually needed here?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "lint": "eslint .",
+    "preview": "vite preview"
+  },
+  "dependencies": {
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
+    "@fontsource/roboto": "^5.1.1",
+    "@mui/icons-material": "^6.4.2",
+    "@mui/material": "^6.4.2",
+    "@tailwindcss/vite": "^4.0.0",
```
Comment (padelis) [inline package.json:19]: How come you chose both `mui` and `tailwind` together?

**Review by padelis — COMMENTED** [review]
Hello @OrestisZinelis! 👋 
First of all, thanks for taking the time and effort to do the assignment! 

I've added a couple of questions if you wanna take a look. 
Also have you used any tools to format your code?

File: src/App.tsx (hunk 1 above)
Comment (OrestisZinelis) [reply src/App.tsx]: You are right, it should not be included in the dependencies array. It will never change across renders, because it is from `useState`.

**Comment by OrestisZinelis** [comment]
> Hello @OrestisZinelis! 👋 First of all, thanks for taking the time and effort to do the assignment!
> 
> I've added a couple of questions if you wanna take a look. Also have you used any tools to format your code?

I have used Prettier (via VSCode extension) but I did not install prettier and configure pretier with eslint...  I added a commit

File: package.json (hunk 2 above)
Comment (OrestisZinelis) [reply package.json:19]: I started the project using tailwind and then took the decision to use material UI 😅 . 

In some places I could use the sx props or theme override. I could have done a better job at this, be more consistent.

**Comment by padelis** [comment]
Thanks again for the prompt replies and fixes

File: package.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    "dev": "vite",
+    "build": "tsc -b && vite build",
+    "lint": "eslint .",
+    "format": "prettier . --write",
+    "preview": "vite preview"
+  },
+  "dependencies": {
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
+    "@fontsource/roboto": "^5.1.1",
```
Comment (gpositive) [inline package.json:16]: Is this font-family actually needed? I can't see this font being used anywhere but I might be wrong.

File: src/pages/Cats.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        {isFetching ? (
+          <SkeletonImages />
+        ) : (
+          (cats || []).map((cat) => (
+            <ImageListItem key={cat.id}>
+              <Link to={`${cat.id}`}>
+                <img src={cat.url} alt="Cat" loading="lazy" />
+                <ImageListItemBar
+                  position="bottom"
+                  title={cat?.breeds[0]?.name ?? ""}
```
Comment (gpositive) [inline src/pages/Cats.tsx]: There was a case where I got a white screen of death that was coming from this line.
Unfortunately I haven't managed to debug in depth.
Can you guess why this might have happened?

File: src/components/BreedModal.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                )}
+              </ImageList>
+
+              <CardActions className="flex justify-center">
+                <Button
+                  variant="contained"
+                  color="secondary"
+                  onClick={() => refetchCats()}
+                >
+                  Load more
```
Comment (gpositive) [inline src/components/BreedModal.tsx:97]: Correct me if I'm wrong but right now, clicking 'Load More' fetches the same data repeatedly. Is this intended behaviour? If not are there any ways to improve it?

File: src/api/favorite.api.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import type { Cat } from "../types/api/Cat.types";
+
+export const addFavorite = async (
+  imageId: Cat["id"],
+): Promise<{ id: number; imageUrl: string }> => {
+  const { data } = await api.post<{ id: number }>("/favourites", {
+    image_id: imageId,
+  });
+
+  const cat = await getCatById(imageId); // Fetch the image URL associated with this imageId to avoid refetching favorites
```
Comment (gpositive) [inline src/api/favorite.api.ts:13]: Could you please explain this a bit more?

File: src/components/BreedModal.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+              <p className="text-center text-gray-600">{breed?.description}</p>
+
+              <ImageList className="mt-4" variant="masonry" cols={cols} gap={8}>
+                {isFetchingCats ? (
+                  <SkeletonImages height="5vh" />
+                ) : (
+                  (cats || []).map((cat) => (
+                    <ImageListItem key={cat.id}>
+                      <Link to={`./cats/${cat.id}`} relative="path">
+                        <img src={cat.url} alt="Cat" loading="lazy" />
```
Comment (gpositive) [inline src/components/BreedModal.tsx]: Can you explain what loading="lazy" does and why you chose to use it here?

File: src/hooks/queries/cat.queries.use.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    queryKey: ["cat", id],
+    queryFn: () => getCatById(id),
+    retry: false,
+  });
+
+export const useGetCatsByBreedId = (breedId: string) =>
+  useQuery({
+    queryKey: ["cats-by-breed", breedId],
+    queryFn: () => getCatsByBreedId(breedId, 10),
+    refetchOnMount: false,
```
Comment (gpositive) [inline src/hooks/queries/cat.queries.use.ts:25]: In some of these `useQuery` params you're setting `refetchOnMount: false` but not in others. What are the reasons behind this decision?

File: src/layouts/AppTopBar.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          <Typography
+            variant="h6"
+            noWrap
+            component="div"
+            sx={{
+              mr: 2,
+              display: { xs: "none", md: "flex" },
+              fontFamily: "monospace",
+              fontWeight: 700,
+              letterSpacing: ".3rem",
```
Comment (gpositive) [inline src/layouts/AppTopBar.tsx:64]: What's the difference between `rem` and `em`? Why you've chosen to go with `rem` in this specific example?

**Review by gpositive — COMMENTED** [review]
Heya Oresti! 👋

Thank you for taking the time to complete this assignment! We really appreciate it!

I've already added some comments and some more questions for you to answer if you can.

- Is there anything you would improve in this assignment if you had more time?
- Testing is missing from your project. If you were to test your application, how would you approach this?

Thanks again!

File: package.json (hunk 3 above)
Comment (OrestisZinelis) [reply package.json:16]:
```
<style data-emotion="css-global" data-s="">body{margin:0;color:rgba(0, 0, 0, 0.87);font-family:"Roboto","Helvetica","Arial", sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;background-color:#fff;}</style>
```

Material UI applies it. I followed the instructions about the Roboto font from the Material UI documentation. The font is used but is not visible from the code source.

File: src/layouts/AppTopBar.tsx (hunk 9 above)
Comment (OrestisZinelis) [reply src/layouts/AppTopBar.tsx:64]: To implement the top bar, I used [this](https://mui.com/material-ui/react-app-bar/?srsltid=AfmBOop09MlaruqQPVysIf0fJA2XDuvERMlk7Xw0-GGGFi1zVUnUodpx#system-ResponsiveAppBar.tsx) example. I did not check the code thoroughly. 

In general, the root element (rem) is relative to the root <HTML>. em is relative to the font size of the parent element.

I guess that for a fixed, consistent look like a logo or branding element, rem is the right choice.

File: src/components/BreedModal.tsx (hunk 5 above)
Comment (OrestisZinelis) [reply src/components/BreedModal.tsx:97]: This is a tricky one. It depends on the breed. If there are more than 10 images available for a breed, then the user may get different images on refetch (e.g. Maine Coon breed).

File: src/pages/Cats.tsx (hunk 4 above)
Comment (OrestisZinelis) [reply src/pages/Cats.tsx]: I cannot reproduce it, unfortunately.

In the api request I use `has_breeds=1`, I get only cats with available breed, but needs a valid API key. The main problem is that I do not handle any possible fetch error. I initialize the cats as an empty array, but the code will crash if `breeds` is not an array. 

I prepared a commit to handle fetching errors or empty data.

File: src/api/favorite.api.ts (hunk 6 above)
Comment (OrestisZinelis) [reply src/api/favorite.api.ts:13]: Each time I add a favorite, I also fetch the cat to update the imageUrl in `useAddFavoriteMutation`.

```ts
 return oldFavorites.map((fav) =>
            fav.image_id === imageId
              ? {
                  ...fav,
                  id: data.id,
                  image_id: imageId,
                  image: { id: imageId, url: data.imageUrl },
                }
              : fav,
          );
```
          
This way, the imageUrl will exist for an added favorite and the favorites data will be consistent, without the need to fetch favorites (update oldFavorites array optimistically). 

Pros: favorites data is not refetched each time we access the favorites page
Cons: An additional get request each time user adds a favorite.

File: src/components/BreedModal.tsx (hunk 7 above)
Comment (OrestisZinelis) [reply src/components/BreedModal.tsx]: `lazy` will delay the loading of the image until it is about to scroll into view. I should not have used it here. If we fetch many images, that are not visible to the user then `lazy` should be used.

File: src/hooks/queries/cat.queries.use.ts (hunk 8 above)
Comment (OrestisZinelis) [reply src/hooks/queries/cat.queries.use.ts:25]: I use everywhere `refetchOnMount: false` except  in `useGetCatById` query which is a mistake (I missed it)

For Cats page: I did not want to refetch random cats, only onClick load more button.
For Breeds page: (table) I did not want to refetch data. Breeds data won't change so often.
For Favorites page: I update oldFavorites array optimistically, there is no need to fetch favorites on mount

I pushed a fix  for `useGetCatById`

**Comment by OrestisZinelis** [comment]
> Heya Oresti! 👋
> 
> Thank you for taking the time to complete this assignment! We really appreciate it!
> 
> I've already added some comments and some more questions for you to answer if you can.
> 
> * Is there anything you would improve in this assignment if you had more time?
> * Testing is missing from your project. If you were to test your application, how would you approach this?
> 
> Thanks again!

 I would focus on improving the user experience, accessibility, and performance of the application. 

For testing, mocking API responses and testing edge cases would also be a priority to cover all possible scenarios. I would use a combination of unit, integration, and E2E tests.


**Responsive Design:** Now the app is somewhat responsive for 600px min width.
**Accessibility:** e.g. Add keyboard navigation
**Breeds table:** Search & filter & pagination for breeds
**Better favorites management:** Display some statistics, group by breed, toggle between list/grid view modes, batch deletion
Add **Localization** 
**User Authentication**
